### ISSUE TITLE
Different result between onnx exported pytorch & barracuda

### ISSUE THREAD

**Issue by Sun-Pak** [Unity-Technologies/barracuda-release#172]
HI. something strange found when i run barracuda using onnx model exported from pytorch.
i run the same image between unity and colab, but the result is different. 
I found that the tensor values are different.
but. i use colab tensor value to barracuda, it works well.
I think converting texture2d to tensor is wrong.
The following code referenced the api. 
Please let me know if you have any problems.

    public void SingleImgTest(Texture2D _target)
     {
         var worker = WorkerFactory.CreateWorker(WorkerFactory.Type.ComputePrecompiled, m_RuntimeModel);
         Tensor tensor = new Tensor(_target, 1);  // grayscale

        ITensorData tensorData = tensor.data;

        TensorShape ts = new TensorShape(1, 64, 64, 1);

         var downData = tensorData.Download(ts);  // normalize [0-1]

        string writeData = "";
        for (int i = 0; i < downData.Length; i++)
        {
            downData[i] = 2 * downData[i] - 1;   //normalize [-1, 1]
            writeData += downData[i].ToString() + "\n";
        }

        System.IO.StreamWriter file = new System.IO.StreamWriter("Assets/Resources/test.txt");
        file.WriteLine(writeData);

        Tensor inputTensor = new Tensor(1, 64, 64, 1, downData);

        Tensor result = ExecuteInParts(worker, inputTensor);

        int[] indicies = result.ArgMax();

        resultString = resultTxt[indicies[0]];

        result0 = result.AsFloats();

        print(resultString);

        tensor.Dispose();
        result.Dispose();
        inputTensor.Dispose();
    }

**Comment by FlorentGuinier** [comment]
Hi @Sun-Pak , let me rephrase to be sure i understand the problem:
- The code above is working as expected (Barracuda match colab result), however you would prefer not to have to do the range convertion from [0-1] to [-1,1] manually?

**Comment by Sun-Pak** [comment]
my onnx model trained [-1,1] normalization. so i convert  [0,1] to [-1,1].
same problem happen when using [0,1] in the trained onnx model.
because tensor values are different.
I create a file to see the tensor value .(test.txt)

**Comment by FlorentGuinier** [comment]
Ok, thanks for the clarification. So indeed we are missing flexibility when converting Textures to Tensors. It would be good we provide a scale and bias parameters when converting from texture. 
Thanks for the feedback. This has been added to the backlog.

In the meantime a bunch of potential workaround:
* option1: modify the model in python so it take input in the [0-1] range.
* option2: manually modify the model in Barracuda, by adding a `ScaleBias` layer at the start of the network.
* option3: create intermediate texture in range [-1,1] using a compute shader (or pixel shader+blit) itself using _target as input
* option4: create tensor from texture yourself and apply the convertion as needed.

Regarding option 4 you can take a look at the style transfer demo repo and in particular at `tensorToTextureSRGB`:
CS code:
https://github.com/UnityLabs/barracuda-style-transfer/blob/main/Assets/BarracudaStyleTransfer/BarracudaStyleTransfer.cs
Shader: 
https://github.com/UnityLabs/barracuda-style-transfer/blob/main/Assets/BarracudaStyleTransfer/Resources/TextureToSRGBTensor.compute
The case was different (SRGB to linear convertion) however you can use the same approach.

**Comment by Sun-Pak** [comment]
Thank you for recommand.  option1 is very simple, but i don't think it's a model problem. because, when i compared tensor of [0,1] range ,from same image, in unity and colab , it's different. Therefore, the results in the model trained [0,1] were also different.
I'll try the rest of the options you mentioned.

**Comment by Sun-Pak** [comment]
I have a question related to the above.  
As API, Using "Texture as input", I can use this code
var tensor = new Tensor(texture, channelCount);
https://docs.unity3d.com/Packages/com.unity.barracuda@1.0/manual/TensorHandling.html

when i used this code, the tensor of 1channel image was different from colab. 
Here, the curious question is how the bias and scale affect when converting a 1-channel image into a tensor.
I also tried converting the pixel value of the image to a tensor but I got the same different result.
Usually, it has an error of about 0.1~0.01.

**Comment by FlorentGuinier** [comment]
> Thank you for recommand. option1 is very simple, but i don't think it's a model problem. because, when i compared tensor of [0,1] range ,from same image, in unity and colab , it's different. Therefore, the results in the model trained [0,1] were also different.

You might be encoutering a color space different between python input texture and C# texture. I would dig toward that direction (dump values and verify you actually have exact same data as Tensor input to network).

**Comment by Sun-Pak** [comment]
well, I made a simple exercise to help you.( different tensor result between colab and unity) 
It is easy to understand if you compare the text file output from colab and unity.
I agree with you that the color space is different. I think the way Unity reads images is different from Python.
+ The axe image has a different tensor result. but simple straight line image has same tensor.
https://colab.research.google.com/drive/1xeUjnRIaU1EnUHgY0egtPJx_jKA9WwF9?usp=sharing

**Comment by FlorentGuinier** [comment]
> I have a question related to the above.
> As API, Using "Texture as input", I can use this code
> var tensor = new Tensor(texture, channelCount);
> https://docs.unity3d.com/Packages/com.unity.barracuda@1.0/manual/TensorHandling.html
> when i used this code, the tensor of 1channel image was different from colab.
> Here, the curious question is how the bias and scale affect when converting a 1-channel image into a tensor.
> I also tried converting the pixel value of the image to a tensor but I got the same different result.
> Usually, it has an error of about 0.1~0.01.

the channelCount parameter allow to override default behavior when converting texture to tensor. By default you have:
`var tensor = new Tensor(texture);`
tensor will have a shape of (1,texture.height,texture.width,texture.numchannels) for example 4 for an RGBA texture.

however if `var tensor = new Tensor(texture,1);` is called
tensor will have a shape of (1,texture.height,texture.width,1), if texture is one channel only it will be fill with this data (R or A texture). In the case of texture with more than one channel Barracuda will set the tensor with an average or the RGB values.

**Comment by Sun-Pak** [comment]
Yes,  in my case, i use 1-channel image. so It is correct to use.
Tensor tensor = new Tensor(_target, 1);

**Comment by FlorentGuinier** [comment]
Thanks for the code. I played a bit with it and try to compare tensor from pytorch and unity. To do that i have used your notebook and added the following:

print("shape1:",img_tensor.shape)
print("type1:",type(img_tensor))
print("std", img_tensor.std())
print("mean", img_tensor.mean())
print("sum", img_tensor.sum())``

pytorch tensor std,mean and sum is then
std tensor(0.1535)
mean tensor(0.0256)
sum tensor(104.6628)

Then i have match it to Unity using the following:

    public class SingleImgTest : MonoBehaviour
    {

    public Texture2D _target;
    public float stdTex = 0.0f;
    public float meanTex = 0.0f;
    public float sumTex = 0.0f;
    public float stdTensor = 0.0f;
    public float meanTensor = 0.0f;
    public float sumTensor = 0.0f;

    // Start is called before the first frame update
    void Start()
    {
        //Texture를 tensor로 전환시, 데이터는 0~1로 정규화 된다.
        Tensor tensor = new Tensor(_target, 1);

        //tensor data에서 single[] data셋을 뽑아내는 과정
        //ItensorData 인터페이스 rawdata을 추출
        ITensorData tensorData = tensor.data;

        TensorShape ts = new TensorShape(1, 64, 64, 1);

        //Download()의 파라미터는 TensorShape 으로 Input과 일치시켜준다.
        //Download의 return 값은 single[] //float[]와 동일
        var downData = tensorData.Download(ts);  //[0-1]

        //Barracuda Tensor mean and std
        meanTensor = downData.Average();
        stdTensor = GetStd(downData);
        sumTensor = downData.Sum();

        //Texture mean and std
        var pixels = _target.GetPixels();
        float[] rpixels = pixels.Select(val => val.r).ToArray();
        meanTex = rpixels.Average();
        stdTex = GetStd(rpixels);
        sumTex = rpixels.Sum();
    }

    // Update is called once per frame
    void Update()
    {

    }

    float GetStd(float[] values)
    {
        float average = values.Average();
        float sumOfSquaresOfDifferences = values.Select(val => (val - average) * (val - average)).Sum();
        return Mathf.Sqrt(sumOfSquaresOfDifferences / values.Length);
    }
    }

I can see that the value match with the following import settings (test project being in linear color mode https://docs.unity3d.com/Manual/LinearRendering-LinearOrGammaWorkflow.html) :
![image](https://user-images.githubusercontent.com/20201082/117427058-e922fd80-af24-11eb-9427-5ab6ac43f92d.png)

My advice would be to try to repro while validating both project color mode and texture import setting.

**Comment by Sun-Pak** [comment]
Oh. God! Thank you! 
It works well. You're correct to matching texture import setting.
Thank you for solving the problem for a long time together.
Have a nice day!
